queue/driver_mongo.go

```diff
@@ -1260,58 +1250,139 @@ func (d *mongoDriver) scopesInUse(ctx context.Context, scopes []string) bool {
 func (d *mongoDriver) Stats(ctx context.Context) amboy.QueueStats {
 	coll := d.getCollection()
 
-	var numJobs int64
-	var err error
-	if d.opts.UseGroups {
-		numJobs, err = coll.CountDocuments(ctx, bson.M{"group": d.opts.GroupName})
-	} else {
-		numJobs, err = coll.EstimatedDocumentCount(ctx)
+	statusFilter := bson.M{
+		"$or": []bson.M{
+			d.getPendingQuery(bson.M{}),
+			d.getInProgQuery(bson.M{}),
+			d.getRetryingQuery(bson.M{}),
+		},
 	}
+	d.modifyQueryForGroup(statusFilter)
+	matchStatus := bson.M{"$match": statusFilter}
+	groupStatuses := bson.M{
+		"$group": bson.M{
+			"_id": bson.M{
+				"completed":   "$status.completed",
+				"in_prog":     "$status.in_prog",
+				"needs_retry": "$retry_info.needs_retry",
+			},
+			"count": bson.M{"$sum": 1},
+		},
+	}
+	pipeline := []bson.M{matchStatus, groupStatuses}
 
-	grip.Warning(message.WrapError(err, message.Fields{
-		"id":         d.instanceID,
-		"service":    "amboy.queue.mdb",
-		"collection": coll.Name(),
-		"operation":  "queue stats",
-		"is_group":   d.opts.UseGroups,
-		"group":      d.opts.GroupName,
-		"message":    "problem counting all jobs",
-	}))
+	c, err := coll.Aggregate(ctx, pipeline)
+	if err != nil {
+		grip.Warning(message.WrapError(err, message.Fields{
+			"id":         d.instanceID,
+			"service":    "amboy.queue.mdb",
+			"collection": coll.Name(),
+			"is_group":   d.opts.UseGroups,
+			"group":      d.opts.GroupName,
+			"operation":  "queue stats",
+			"message":    "could not count documents by status",
+		}))
+		return amboy.QueueStats{}
+	}
+	statusGroups := []struct {
+		ID struct {
+			Completed  bool `bson:"completed"`
+			InProg     bool `bson:"in_prog"`
+			NeedsRetry bool `bson:"needs_retry"`
+		} `bson:"_id"`
+		Count int `bson:"count"`
+	}{}
+	if err := c.All(ctx, &statusGroups); err != nil {
+		grip.Warning(message.WrapError(err, message.Fields{
+			"id":         d.instanceID,
+			"service":    "amboy.queue.mdb",
+			"collection": coll.Name(),
+			"is_group":   d.opts.UseGroups,
+			"group":      d.opts.GroupName,
+			"operation":  "queue stats",
+			"message":    "failed to decode counts by status",
+		}))
+		return amboy.QueueStats{}
+	}
 
-	pendingQuery := bson.M{"status.completed": false}
-	d.modifyQueryForGroup(pendingQuery)
-	pending, err := coll.CountDocuments(ctx, pendingQuery)
+	var pending, inProg, retrying int
+	for _, group := range statusGroups {
+		if !group.ID.Completed {
+			pending += group.Count
+		}
+		if group.ID.InProg {
+			inProg += group.Count
+		}
+		if group.ID.Completed && group.ID.NeedsRetry {
+			retrying += group.Count
+		}
+	}
+
+	// The aggregation cannot also count all the documents in the collection
```

Comment:
Could we use a transaction instead so we can do this completed count atomically?

Reply:
I think this could be risky since the Amboy stats job relies on this to be fast and there were some issues with performance with using transactions in Evergreen. I think it's okay for it to be slightly wrong because:
1. The two queries are issued within a short time frame.
2. They're not critical, since they're mostly used for tests (which don't care too deeply about atomicity of stats) and Splunk stats (which don't care that much about # completed jobs).